### PR TITLE
tools: add comprehensive coverage reporting to router check

### PR DESCRIPTION
**Pull request by dschaller**
Description: add comprehensive coverage reporting to the router check tool.
Risk Level: Low
Testing: Added tests and manually tested
Docs Changes: Included
Release Notes: Included

### CODE REVIEW

**Comment by dschaller** [comment]
@dio @rgs1 @mattklein123 @jyotima mind taking a look at this? It’s an expansion off of https://github.com/envoyproxy/envoy/pull/7727

File: test/tools/router_check/coverage.h [hunk 1]
```diff
@@ -5,14 +5,52 @@
...
+  void setClusterCovered() { cluster_covered_ = true; }
+  void setVirtualClusterCovered() { virtual_cluster_covered_ = true; }
+  void setVirtualHostCovered() { virtual_host_covered_ = true; }
+  void setPathRewriteCovered() { path_rewrite_covered_ = true; }
+  void setHostRewriteCovered() { host_rewrite_covered_ = true; }
+  void setRedirectPathCovered() { redirect_path_covered_ = true; }
+  bool covers(const Envoy::Router::RouteEntry* route) { return route_ == route; }
+
+private:
+  const Envoy::Router::RouteEntry* route_;
```
Comment (mattklein123) [inline test/tools/router_check/coverage.h]: nit: this can be a reference

File: test/tools/router_check/coverage.h [hunk 2]
```diff
@@ -5,14 +5,52 @@
...
+  void markPathRewriteCovered(const Envoy::Router::RouteEntry& route);
+  void markHostRewriteCovered(const Envoy::Router::RouteEntry& route);
+  void markRedirectPathCovered(const Envoy::Router::RouteEntry& route);
   double report();
+  double detailedReport();
+  void cleanup();
 
 private:
-  std::vector<const Envoy::Router::RouteEntry*> seen_routes_;
+  RouteCoverage* coveredRoute(const Envoy::Router::RouteEntry& route);
```
Comment (mattklein123) [inline test/tools/router_check/coverage.h]: nit: return reference

File: test/tools/router_check/router.h [hunk 3]
```diff
@@ -89,7 +89,11 @@ class RouterCheckTool : Logger::Loggable<Logger::Id::testing> {
    */
   void setShowDetails() { details_ = true; }
 
-  float coverage() { return coverage_.report(); }
+  float coverage(const bool detailed) {
```
Comment (mattklein123) [inline test/tools/router_check/router.h]: nit: remove const, doesn't do anything

File: test/tools/router_check/coverage.cc [hunk 4]
```diff
@@ -5,19 +5,86 @@
...
+  double cumulative_coverage = 0;
+  for (auto& covered_route : covered_routes_) {
+    cumulative_coverage += covered_route->report();
+  }
+  return 100 * cumulative_coverage / num_routes;
+}
+
+void Coverage::cleanup() {
+  for (auto& covered_route : covered_routes_) {
+    delete covered_route;
```
Comment (mattklein123) [inline test/tools/router_check/coverage.cc]: Store unique_ptr in your list, and then you can remove all of this manual cleanup code. (In general for C++ if you find yourself writing `delete` something is probably wrong.)

**Comment by dschaller** [comment]
Yeah, I was hoping that it would be possible to make this less manual but couldn't think of anything that was promising. I'm especially not a fan of the vector of `coverageFIelds` but couldn't find an alternative.

On the plus side this increases visibility into what fields on routes are covered. 🎉

**Comment by jyotimahapatra** [comment]
#730

**Pull request merged**